homeassistant/components/mqtt/schemas.py

```diff
@@ -37,16 +43,78 @@
     CONF_ORIGIN,
     CONF_PAYLOAD_AVAILABLE,
     CONF_PAYLOAD_NOT_AVAILABLE,
+    CONF_QOS,
     CONF_SERIAL_NUMBER,
+    CONF_STATE_TOPIC,
     CONF_SUGGESTED_AREA,
     CONF_SUPPORT_URL,
     CONF_SW_VERSION,
     CONF_TOPIC,
     CONF_VIA_DEVICE,
     DEFAULT_PAYLOAD_AVAILABLE,
     DEFAULT_PAYLOAD_NOT_AVAILABLE,
+    SUPPORTED_COMPONENTS,
+)
+from .util import valid_publish_topic, valid_qos_schema, valid_subscribe_topic
+
+_LOGGER = logging.getLogger(__name__)
+
+# Device discovery options that are also available at entity component level
+SHARED_OPTIONS = [
+    CONF_AVAILABILITY,
+    CONF_AVAILABILITY_MODE,
+    CONF_AVAILABILITY_TEMPLATE,
+    CONF_AVAILABILITY_TOPIC,
+    CONF_COMMAND_TOPIC,
+    CONF_PAYLOAD_AVAILABLE,
+    CONF_PAYLOAD_NOT_AVAILABLE,
+    CONF_STATE_TOPIC,
+]
+
+
+def validate_device_has_at_least_one_identifier(value: ConfigType) -> ConfigType:
+    """Validate that a device info entry has at least one identifying value."""
+    if value.get(CONF_IDENTIFIERS) or value.get(CONF_CONNECTIONS):
+        return value
+    raise vol.Invalid(
+        "Device must have at least one identifying value in "
+        "'identifiers' and/or 'connections'"
+    )
+
```

Comment:
Looks like this was moved. PR would be a bit smaller if its kept in the same place